### ISSUE TITLE
Sinatra leaks sensitive information from any exception with an `http_status` attribute

### ISSUE THREAD

**Issue by garybernhardt** [sinatra/sinatra#1518]
This is fundamentally the same issue as #1204. However, I think I have a clearer explanation of what's happening, including an easily-reproduced example. I also think the bug is a serious security problem and want to make that clear.

First, Stripe's errors have an `http_status` attribute. This attribute represents the status code returned by Stripe's servers. For example, [here](https://github.com/stripe/stripe-ruby/blob/863da48398e3c003556bc4bcde7c165f382ad025/lib/stripe/errors.rb#L82) you can see `Stripe::InvalidRequestError` taking an `http_status` argument, which is 400 when the error is actually raised.

```ruby
  class InvalidRequestError < StripeError
    attr_accessor :param

    def initialize(message, param, http_status: nil, http_body: nil, json_body: nil,
                   http_headers: nil, code: nil)
```

In our application, we were seeing this error raised for correct reasons (i.e., we had an actual error). But Sinatra [will treat](https://github.com/sinatra/sinatra/blob/c686a10be68a959c34495c4969dcbabe7b93205e/lib/sinatra/base.rb#L1116) **any** exception with an `http_status` attribute as a Sinatra-level error that should be translated into that HTTP status code.

```ruby
    def handle_exception!(boom)
      [...]

      if boom.respond_to? :http_status
        status(boom.http_status)
```

Note that the Sinatra code above does not respond to **any** configuration option. As far as I can tell, this problem exists regardless of how Sinatra is configured.

By these two code snippets' powers combined:

1. Stripe's servers are 400ing because we did something wrong.
2. That 400 response exception's text contains sensitive Stripe identifiers, like Stripe token IDs.
3. Sinatra sees the `http_status` attribute and incorrectly thinks that this exception is actually a Sinatra exception.
4. **Sinatra returns that status code and the response text to the user, including sensitive information from Stripe.**

You can reproduce this yourself by installing the "stripe" and "sinatra" gems and running this single-file Sinatra app. Go to `localhost:4567/error_test` and you'll see the "sensitive information from stripe" string leaked to the end user.

```
#!/usr/bin/env ruby

require "sinatra"
require "stripe"

get "/error_test" do
  raise Stripe::InvalidRequestError.new(
    "sensitive information from stripe", 2, http_status: 400)
end
```

**Comment by mwpastore** [comment]
Add an error handler to your app to modify the message.

```ruby
error Stripe::InvalidRequestError do
  "boom"
end
```

**Comment by jkowens** [comment]
Looking at this, it seems the exception message will be printed to the body if it is http status 400 or 404. Still not good. See https://github.com/sinatra/sinatra/blob/26fa5c86220890a02cc0f070ddb2696d74a5b843/lib/sinatra/base.rb#L1132.

I'm wondering why there isn't a `Sinatra::Error` class with internal errors extending from that. Seems like this is a case where checking type would be better than just checking if it quacks.

**Comment by garybernhardt** [comment]
@jkowens I also reprod it with 401 when I was diagnosing it. Didn't dig any deeper into the exact status codes affected, though.

**Comment by garybernhardt** [comment]
@mwpastore We'll work around it now that we know. But the problem is that I've never heard of this in ten years of using Sinatra. And we only discovered it because we happened to see an unhandled `Stripe::InvalidRequestError` during development and tracked it down. We could've just as easily found out because it happened in production, leaking sensitive billing data to the client. Like @jkowens, I wonder why this conditional isn't checking for a specific error class, rather than merely checking for the existence of a method.

**Comment by jkowens** [comment]
It looks like Sinatra has tried to set the `http_status` of an error as the response status back to version 1. The difference in 2.0 is that if `http_status` returns 404 or 400, the message of the error is sent as the body of the response by default. Clearly something that should only be reserved for internal Sinatra errors.

**Comment by namusyaka** [comment]
This is an API design issue, **not a bug**. 
Sinatra provides several features by duck-typing according to the Ruby style.
Thus, I don't intend to take a patch from @jkowens right away (but thanks for working on this issue).

By reporting this issue, I could recognize the problem on the API design.
Indeed, we didn't mention it in our document explicitly.

At least, we can take a following action:
- Towards 3.0.0, we can discuss the best API design.
- Mention the `http_status` method and its duck-typing feature in our document (probably README?).

**Comment by garybernhardt** [comment]
I don't think that mentioning `http_status` in the docs solves the problem. Every Sinatra user would have to manually check for an `http_status` method in every exception class that can be raised by any library in the system.

**Comment by namusyaka** [comment]
@garybernhardt I don't say that mentioning in the document will solve this problem. I said that it's important to clarify the matter until the problem is solved in the future.
Please don't misunderstand

**Comment by garybernhardt** [comment]
@namusyaka Got it, thanks for clarifying and sorry for misunderstanding!

**Comment by SwagDevOps** [comment]
Interesting issue 👍
